src/emulator/storage/upload.ts

```diff
@@ -0,0 +1,188 @@
+
+  /** Resets the state of the UploadService. */
+  public reset(): void {
+    this._uploads = new Map();
+  }
+
+  /**
+   * Handles a multipart file upload which is expected to have the entirety of
+   * the file's contents in a single request.
+   */
+  public multipartUpload(request: MultipartUploadRequest): Upload {
+    const data = Buffer.from(request.dataRaw);
+    const upload = this.initMultipartUpload(request, data.byteLength);
+    this._persistence.deleteFile(upload.path, /* failSilently = */ true);
+    this._persistence.appendBytes(upload.path, data);
+    return upload;
+  }
+
+  private initMultipartUpload(request: MultipartUploadRequest, sizeInBytes: number): Upload {
+    const id = uuidV4();
+    const upload: Upload = {
+      id: uuidV4(),
+      bucketId: request.bucketId,
+      objectId: request.objectId,
+      type: UploadType.MULTIPART,
+      path: this.stagingFileName(id, request.bucketId, request.objectId),
+      status: UploadStatus.FINISHED,
+      metadata: JSON.parse(request.metadataRaw),
+      size: sizeInBytes,
+      authorization: request.authorization,
+    };
+    this._uploads.set(upload.id, upload);
+
+    return upload;
+  }
+
+  /**
+   * Initializes a new ResumableUpload.
+   */
+  public startResumableUpload(request: StartResumableUploadRequest): Upload {
+    const id = uuidV4();
+    const upload: Upload = {
+      id: id,
+      bucketId: request.bucketId,
+      objectId: request.objectId,
+      type: UploadType.RESUMABLE,
+      path: this.stagingFileName(id, request.bucketId, request.objectId),
+      status: UploadStatus.ACTIVE,
+      metadata: JSON.parse(request.metadataRaw),
+      size: 0,
+      authorization: request.authorization,
+    };
+    this._uploads.set(upload.id, upload);
+    this._persistence.deleteFile(upload.path, /* failSilently = */ true);
+    return upload;
+  }
+
+  /**
+   * Appends bytes to an existing resumable upload.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   * @throws {NotActiveUploadError} if the resumable upload is not in the ACTIVE state.
+   */
+  public progressResumableUpload(uploadId: string, dataRaw: string): Upload {
+    const upload = this.findResumableUpload(uploadId);
+    if (upload.status !== UploadStatus.ACTIVE) {
+      throw new UploadNotActiveError();
+    }
+    const data = Buffer.from(dataRaw);
+    this._persistence.appendBytes(upload.path, data);
+    upload.size += data.byteLength;
+    return upload;
+  }
+
+  /**
+   * Queries for an existing resumable upload.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   */
+  public getResumableUpload(uploadId: string): Upload {
+    return this.findResumableUpload(uploadId);
+  }
+
+  /**
+   * Cancels a resumable upload.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   * @throws {NotCancellableError} if the resumable upload can not be cancelled.
+   */
+  public cancelResumableUpload(uploadId: string): Upload {
```

Comment:
I would pass an `Upload` rather than its id here and have the caller worry about `NotFoundError`, similarly for finalise. I've floated this idea [before](https://github.com/firebase/firebase-tools/pull/4210#discussion_r815171916)... Thoughts?